src/platform/packages/shared/kbn-unified-metrics-grid/src/hooks/use_metric_fields.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { useEffect, useMemo, useRef } from 'react';
+import type { DatatableRow } from '@kbn/expressions-plugin/common';
+import type { Dimension, MetricField, MetricUnit } from '../types';
+import { useMetricsExperienceFieldsContext } from '../context/metrics_experience_fields_provider';
+import { normalizeUnit } from '../common/utils/metric_unit/normalize_unit';
+import { hasValue } from '../common/utils/fields';
+import { useMetricsExperienceState } from '../context/metrics_experience_state_provider';
+import { useMetricFieldsFilter } from './use_metric_fields_filter';
+
+interface UseMetricFieldsReturn {
+  allMetricFields: MetricField[];
+  visibleMetricFields: MetricField[];
+  dimensions: Dimension[];
+}
+
+/**
+ * Builds MetricField[] from the context's metric fieldSpecs and sampleRowByMetric.
+ * Returns:
+ * - allMetricFields: Complete set of metric fields (for dimension selector, filtering source)
+ * - visibleMetricFields: Currently visible fields based on filters (for grid, value selector)
+ * - dimensions: Unique dimensions extracted from sampled metric fields
+ */
+export const useMetricFields = (): UseMetricFieldsReturn => {
+  const { searchTerm, selectedDimensions, selectedValuesMetricFields, onDimensionsChange } =
+    useMetricsExperienceState();
+  const { metricFields, dimensions, sampleRowByMetric } = useMetricsExperienceFieldsContext();
+
+  // Ref to access current values in effects without adding them to dependencies
+  const stateRef = useRef({
+    selectedDimensions: [] as Dimension[],
+    returnValue: null as UseMetricFieldsReturn | null,
+  });
+
+  stateRef.current.selectedDimensions = selectedDimensions;
+
+  // Computed values
+  const sampledMetricFields = useMemo(() => {
```

Comment:
I might be missing some context but from reading the code, `sampledMetricFields` makes it sound like this is a sample of the data.

What does it actually represent? cause later on it's return as 

```
 allMetricFields: sampledMetricFields,
```

Reply:
That's all the metric fields that matter for the UI. There are  `allMetricFields`, which are not affected by UI filters, and `visibleMetricFields`, which are.

`allMetricFields`  is derived from the data